api/authenticator.py

```diff
@@ -232,69 +579,704 @@ def create_authentication_headers(self):
 
         return headers
 
-class BasicAuthAuthenticator(Authenticator):
 
-    TYPE = Authenticator.BASIC_AUTH
-    METHOD = "http://opds-spec.org/auth/basic"
-    NAME = _("Library Barcode")
-    URI = "http://librarysimplified.org/terms/auth/library-barcode"
+class AuthenticationProvider(object):
+    """Handle a specific patron authentication scheme.
+    """
+
+    # A subclass MUST define a value for URI. This is used in the
+    # Authentication for OPDS document to distinguish between
+    # different types of authentication.
+    URI = None
+    
+    def authenticated_patron(self, _db, header):
+        """Go from a WWW-Authenticate header (or equivalent) to a Patron object.
+
+        If the Patron needs to have their metadata updated, it happens
+        transparently at this point.
+
+        :return: A Patron if one can be authenticated; a ProblemDetail
+        if an error occurs; None if the credentials are missing or
+        wrong.
+        """
+        patron = self.authenticate(_db, header)
+        if not isinstance(patron, Patron):
+            return patron
+        if patron.needs_external_sync:
+            self.update_patron_metadata(patron)
```

Comment:
That's true. I'm having trouble thinking of a good solution. The best I can come up with is to set the sync timeout for patrons without borrowing privileges to, like, five seconds, instead of zero.